site/app/controllers/grading/ElectronicGraderController.php

```diff
@@ -48,6 +52,101 @@ private function validateDiffViewerType(string $type) {
+        3 - Randomize the order of students
+        4 - Randomly Select Y offsets
+        5 - Shift the random order by the offsets to create the matrix, with no duplicates, and exactly Y assignments and & graders for each student.  no student grades self.
+        */
+        $number_to_grade = $_POST['number_to_grade'];
+        $gradeable = $this->tryGetGradeable($gradeable_id);
+        if ($gradeable === false) {
+            $this->core->addErrorMessage('Invalid Gradeable!');
+            $this->core->redirect($this->core->buildCourseUrl());
+        }
+        $all_grade_all = false;
+        $order = new GradingOrder($this->core, $gradeable, $this->core->getUser(), true);
+        $student_array = [];
+        $student_list = [];
+        $students = $this->core->getQueries()->getUsersByRegistrationSections($order->getSectionNames());
+        foreach ($students as $student) {
+             $reg_sec = ($student->getRegistrationSection() === null) ? 'NULL' : $student->getRegistrationSection();
+             $sorted_students[$reg_sec][] = $student;
+             array_push($student_list, ['user_id' => $student->getId()]);
+             array_push($student_array, $student->getId());
+        }
+        $number_of_students = count($student_list);
+        if ($number_to_grade > $number_of_students) {
+            $all_grade_all = true;
+        }
+        if ($all_grade_all) {
+            $this->core->getOutput()->renderJSONSuccess("Invalid Number of Students Entered");
+            $final_grading_info = [];
+            for ($grader = 0; $grader < count($student_array); ++$grader) {
+                $peer_array = [];
+                for ($peer = 0; $peer < count($student_array); ++$peer) {
+                    if ($grader === $peer) {
+                        continue;
+                    }
+                    else {
+                        array_push($peer_array, $student_array[$peer]);
+                    }
+                }
+                array_push($final_grading_info, [$student_array[$grader],$peer_array]);
+            }
+            $gradeable->setRandomPeerGradersList($final_grading_info);
+            return;
+        }
+        $graded_array = $student_array;
+          /*n_array_peers : An Array of arrays that holds information on to be graded peers
+          [ [A,B,C,D,E,F], [E,F,A,B,C,D], [C,D,E,F,A,B] ]
+          A grades C and E and is graded by C and E.
+          */
+        $n_array_peers = [];
+        shuffle($student_array);
+        array_push($n_array_peers, $student_array);
+          /*final_grading_info : An Array with clear structure of grading rules for peer grading
+          [ [A,[C,E]],[B,[F,D]], ...]
+          A grades C and E, B grades F and D ..and so on!
+          */
+        $final_grading_info = [];
+        $max_offset = count($student_array);
+        $offset_array = [];
+        for ($i = 0; $i < $number_to_grade; ++$i) {
+            $random_offset = rand(1, $max_offset);
```

Comment:
For this to work, we can't ever get duplicate offsets. 

For reference, here's the algorithm @bmcutler wrote on slack:
```
Example where X=7 and Y=3
7 students:  a b c d e f g
The algorithm we would like you to implement:
1. randomize the order of the students
   c f d b g a e
2. randomly select Y offsets
   from the set { 1, 2, 3, 4, 5, 6 } with no duplicates
   for example, we might select { 1, 4, 5 }
3. shift the random order by the offsets to create the matrix,
   with no duplicates, and exactly Y assignments and & graders for
   each student.  no student grades self.
   student             c f d b g a e
   grader (offset=1)   e c f d b g a
   grader (offset=4)   b g a e c f d
   grader (offset=5)   d b g a e c f
4. So in this example, student g grades peers b, c, e and receives
   grades from f, d, and a.
```

Reply:
Corrected!